massa-execution-worker/src/speculative_deferred_calls.rs

```diff
@@ -0,0 +1,517 @@
+                    .state_changes
+                    .deferred_call_changes
+                    .get_call(&slot, id)
+                {
+                    return Some(v.clone());
+                }
+            }
+        }
+
+        // check final state
+        {
+            let final_state = self.final_state.read();
+            // if let Some(v) = final_state.get_deferred_call_registry().get_call(&slot, id) {
+            if let Some(v) = final_state.get_deferred_call_registry().get_call(&slot, id) {
+                return Some(v.clone());
+            }
+        }
+
+        None
+    }
+
+    pub fn delete_call(&mut self, id: &DeferredCallId, slot: Slot) {
+        self.deferred_calls_changes.delete_call(slot, id)
+    }
+
+    /// Cancel a call
+    /// Returns the sender address and the amount of coins to reimburse them
+    pub fn cancel_call(
+        &mut self,
+        id: &DeferredCallId,
+    ) -> Result<(Address, Amount), ExecutionError> {
+        // get call, fail if it does not exist
+        let Some(mut call) = self.get_call(id) else {
+            return Err(ExecutionError::DeferredCallsError(
+                "Call ID does not exist.".into(),
+            ));
+        };
+
+        // check if the call is already cancelled
+        if call.cancelled {
+            return Err(ExecutionError::DeferredCallsError(
+                "Call ID is already cancelled.".into(),
+            ));
+        }
+
+        // set call as cancelled
+        call.cancelled = true;
+
+        // we need to reimburse coins to the sender
+        let res = (call.sender_address, call.coins);
+
+        // Add a cancellation to the current changes
+        self.deferred_calls_changes
+            .set_call(id.clone(), call.clone());
+
+        let current_gas = self.get_slot_gas(&call.target_slot);
+
+        // set slot gas
+        self.deferred_calls_changes
+            .set_slot_gas(call.target_slot, current_gas - call.max_gas);
+
+        // set total gas
+        self.deferred_calls_changes
+            .set_total_gas(self.get_total_gas().saturating_sub(call.max_gas as u128));
+
+        Ok(res)
+    }
+
+    // This function assumes that we have a resource with a total supply `resource_supply`.
+    // Below a certain target occupancy `target_occupancy` of that resource, the overbooking penalty for using a unit of the resource is zero.
+    // Above the target, the resource unit cost grows linearly with occupancy.
+    // The linear fee growth is chosen so that if someone occupies all the resource, they incur a `max_penalty` cost.
+    fn overbooking_fee(
+        resource_supply: u128,
+        target_occupancy: u128,
+        current_occupancy: u128,
+        resource_request: u128,
+        max_penalty: Amount,
+    ) -> Amount {
+        // linear part of the occupancy before booking the requested amount
+        let relu_occupancy_before =
+            std::cmp::max(current_occupancy, target_occupancy) - target_occupancy;
```

Comment:
Why not directly use `current_occupancy.saturating_sub(target_occupancy)`?

Reply:
done